lib/api_remote.ml

```diff
@@ -225,6 +230,64 @@ module Slack : Api.Slack = struct
         (Option.default "" res.error);
       Lwt.return_none
     | Ok ({ permalink; _ } : Slack_t.permalink_res) -> Lwt.return_some permalink
+
+  let get_upload_URL_external ~ctx ~filename ~size =
+    let url_args = Web.make_url_args [ "filename", filename; "length", Int.to_string size ] in
+    request_token_auth ~name:"get_upload_URL_external" ~ctx `GET
+      (sprintf "files.getUploadURLExternal?%s" url_args)
+      Slack_j.read_upload_url_res
+
+  let post_file_content ~upload_url ~filename ~content =
+    match%lwt http_request ~body:(`Raw ("text/plain", content)) `POST upload_url with
+    | Error e -> Lwt.return_error e
+    | Ok s ->
+      log#info "uploaded file for %s. Reply: %s" filename s;
+      Lwt.return_ok ()
+
+  let pre_upload ~ctx (file : Slack.file) =
+    let { Slack.name; alt_txt = _; content; title } = file in
+    match%lwt get_upload_URL_external ~ctx ~filename:name ~size:(String.length content) with
+    | Error e -> Lwt.return_error e
+    | Ok { Slack_t.upload_url; file_id } ->
+      (match%lwt post_file_content ~upload_url ~filename:name ~content with
+      | Error e -> Lwt.return_error e
+      | Ok () -> Lwt.return_ok ({ id = file_id; title } : Slack_t.file_v2))
+
+  let rec complete_upload_external ~(ctx : Context.t) ?(retry = false) ?channel_id ?thread_ts ?initial_comment files =
+    log#info "send file: complete_upload_external to channel: %S"
+      (match channel_id with
+      | None -> "None"
+      | Some c -> Slack_channel.Ident.project c);
+    let req = { Slack_t.files; channel_id; thread_ts; initial_comment } in
+    let data = Slack_j.string_of_complete_upload_external_req req in
+    log#info "data: %s" data;
+    let body = `Raw ("application/json; charset=utf-8", data) in
+    match%lwt
+      Lwt_result.bind
+        (request_token_auth ~name:"complete_upload_external" ~ctx `POST ~body
+           (sprintf "files.completeUploadExternal")
+           Slack_j.read_complete_upload_external_res)
+        (fun _res -> Lwt.return_ok ())
+    with
+    | Error e when (not retry) && String.ends_with ~suffix:"not_in_channel" e ->
+      let channel = Option.get channel_id in
+      (match%lwt join_channel ~ctx channel with
+      | Error e -> Lwt.return_error e
+      | Ok _ -> complete_upload_external ~ctx ~retry:true ?channel_id ?thread_ts ?initial_comment files)
```

Comment:
1 - can you turn the not_in_channel error into a variant? Also `"not_in_channel"` is the whole text of the error, not the suffix

2 - When you have function that calls itself at most once, that's not a recursive function. you can just repeat the request above to `files.completeUploadExternal` and have that (probably) be the return value of the `complete_upload_external` function.

Reply:
> 1 - can you turn the not_in_channel error into a variant

No, because `request_token_auth` assumes errors are strings.

> Also "not_in_channel" is the whole text of the error, not the suffix

Its not, `request_token_auth` adds stuff to it.

> When you have function that calls itself at most once, that's not a recursive function. you can just repeat the request above to files.completeUploadExternal and have that (probably) be the return value of the complete_upload_external function.

This is done.